Source/MQTTnet/Server/MqttRetainedMessagesManager.cs

```diff
@@ -107,6 +116,11 @@ public async Task<List<MqttApplicationMessage>> GetSubscribedMessagesAsync(IColl
         {
             if (topicFilters == null) throw new ArgumentNullException(nameof(topicFilters));
 
+            if (_options.Storage is IMqttExtendedServerStorage extendedStorage)
+            {
+                return await extendedStorage.GetSubscribedMessagesAsync(topicFilters).ConfigureAwait(false);
```

Comment:
The short answers is yes, but using a different implementation.

We want to use Redis as our retained messages backend (partially because we already use it as the application bus, but also because we need to access the same data from other services that don't speak MQTT). While we could retrieve all the messages on start and then keep the dictionary in `MqttRetainedMessagesManager` in sync with Redis on write, we are talking about tens of thousands of keys and, given that each client only needs 10-100 of them, we better like to keep them in Redis and return them on request. We plant to convert the filters to a single regular expression and use that to retrieve the correct messages from Redis.

I am more than open to discuss and implement a better interface if you have any idea.

Also, if you are interested I can provide a "toy" implementation that store retained messages as separate files on the file system and the real Redis backend we are developing. Just tell me where to put the files.

Reply:
OK I got the point. Writing the retained messages in single files is also interesting for me. At the moment I have my custom storage implementation which will write all message to a storage. I also tried to identify changes in the complete list and only writing new/updated ones. The key I use is the topic. We can also try to add new methods to the existing interface. I was thinking about having 1 method which gives you all messages and additional ones which are called for a delete and a update/add. Then everyone can choose from that methods because all care called from the lib (update/add/remove first and "changed" after the other ones).

But this idea will not solve the reading issues you have at the moment. Please give me some days to think about that. The problem is that by MQTT spec the broker "owns" the retained messages. That is how it works now. We can also add some more abstraction layers so that there is a new service (class) which is responsible for delivering the retained messages. It is something NEW, not an extension of the storage interface. The storage interface stays as it is but you get a new interface for your needs. The default implementation works as usual and makes use of the storage. The default implementation is also responsible to keep all retained messages in memory. So to make it short: The "MqttRetainedMessagesManager" gets a proper interface and can be injected via server options. What do you think?

Reply:
Having the ability to inject the `MqttRetainedMessagesManager` is the perfect solution, at least for us. I'll explain why first and discuss how later.

In our specific use case keeping all the retained messages in memory is just overhead. When a new client connects and we must retrieve the messages we do a **single** Redis query and we gather all the messages, ready for delivery. We didn't do a benchmark but, given that the topic filter selects a very small part of the retained key space (something like 1 in 10000) we are almost as fast a iterating over a very big dictionary to apply the topic filters.

Also, we don't like to have 2 different sources of truth. For us, the authority for the remote device status is Redis and we don't like the idea to have that state replicated in a dictionary in the MQTT server. If the two would diverge (for a bug or any other reason) there would be trouble.

During the next few days, if you agree, I could try to implement this:

1) An `IMqttRetainedMessagesManager` interface, that initially mimic the current `MqttRetainedMessagesManager` but can evolve as you see fit.
2) A new server option, to inject the retained message manager.
3) Possibly expand the scope of the `IMqttServerStorage` to include update/delete functionality (to avoid massive data dumps every time a new message is received).

I could also provide example code to show how to use the new features (just tell me how to organize it).